tests/common/dualtor/tunnel_traffic_utils.py

```diff
@@ -0,0 +1,131 @@
+"""Tunnel traffic verification utilities."""
+import ipaddress
+import logging
+import operator
+import pytest
+import sys
+
+from io import BytesIO
+from ptf import mask, testutils
+from scapy.all import IP, Ether
+from tests.common.dualtor import dual_tor_utils
+
+
+@pytest.fixture(scope="function")
+def tunnel_traffic_monitor(ptfadapter, tbinfo):
+
+    class TunnelTrafficMonitor(object):
+        """Monit tunnel traffic from standby ToR to active ToR."""
+
+        @staticmethod
+        def _get_t1_ptf_port_indexes(dut, tbinfo):
+            """Get the port indexes of those ptf port connecting to T1 switches."""
+            pc_ports = dual_tor_utils.get_t1_ptf_pc_ports(dut, tbinfo)
+            return [int(_.strip("eth")) for _ in reduce(operator.add, pc_ports.values(), [])]
+
+        @staticmethod
+        def _find_ipv4_lo_addr(config_facts):
+            """Find the ipv4 Loopback0 address."""
+            for addr in config_facts["LOOPBACK_INTERFACE"]["Loopback0"]:
+                if isinstance(ipaddress.ip_network(addr), ipaddress.IPv4Network):
+                    return addr.split("/")[0]
+
+        @staticmethod
+        def _build_tunnel_packet(out_src_ip, out_dst_ip):
+            """Build the expected tunnel packet."""
+            exp_pkt = testutils.simple_ip_packet(
+                ip_src=out_src_ip,
+                ip_dst=out_dst_ip,
+                pktlen=20
+            )
+            exp_pkt = mask.Mask(exp_pkt)
+            exp_pkt.set_do_not_care_scapy(Ether, "dst")
+            exp_pkt.set_do_not_care_scapy(Ether, "src")
+            exp_pkt.set_do_not_care_scapy(IP, "ihl")
+            exp_pkt.set_do_not_care_scapy(IP, "tos")
+            exp_pkt.set_do_not_care_scapy(IP, "len")
+            exp_pkt.set_do_not_care_scapy(IP, "id")
+            exp_pkt.set_do_not_care_scapy(IP, "flags")
+            exp_pkt.set_do_not_care_scapy(IP, "frag")
+            exp_pkt.set_do_not_care_scapy(IP, "ttl")
+            exp_pkt.set_do_not_care_scapy(IP, "proto")
+            exp_pkt.set_do_not_care_scapy(IP, "chksum")
+            exp_pkt.set_ignore_extra_bytes()
+            return exp_pkt
+
+        @staticmethod
+        def _dump_show_str(packet):
+            """Dump packet show output to string."""
+            _stdout, sys.stdout = sys.stdout, BytesIO()
+            try:
+                packet.show()
+                return sys.stdout.getvalue()
+            finally:
+                sys.stdout = _stdout
+
+        @staticmethod
+        def _check_ttl(packet):
+            """Check ttl field in the packet."""
+            outer_ttl, inner_ttl = packet[IP].ttl, packet[IP].payload[IP].ttl
+            logging.debug("Outer packet TTL: %s, inner packet TTL: %s", outer_ttl, inner_ttl)
+            if outer_ttl != inner_ttl:
```

Comment:
Oh I thought the inner packet TTL is decremented by 1 compared to outer TTL. Can you confirm that they both should be equal?

Reply:
This packet is dumped out of the intercepted ptf port, in this case, it just leaves the standby ToR and hasn't entered the T1 switch yet, so it should have the same TTL as the inner packet.

Reply:
TTL mode is defined as "PIPE", so outer_ttl is expected to be fixed to 255 and inner_ttl un-modified.

Reply:
OK, thanks for the details, will update then. @prsunny